src/main/java/webserver/Controller.java

```diff
@@ -21,20 +34,87 @@ public ResponseEntity service(final RequestEntity request) {
+            return ResponseEntity.redirectResponseEntity("/index.html");
+        }
 
         try {
-            byte[] file = fileService.serveFile(request.getHeader());
-            return ResponseEntity.of(request.getHeader().getPath(), file);
+            return getResponseFile(request);
         } catch (Exception e) {
             throw new IllegalArgumentException("파일이 존재하지 않습니다.");
         }
     }
 
+    private ResponseEntity getResponseFile(final RequestEntity request) throws IOException, URISyntaxException {
+        if (request.pathEquals("/user/list.html")) {
+            return getUserList(request);
+        }
+        // fallback
+        if (request.getPath().endsWith(".html")) {
+            final byte[] file = fileService.render(request.getPath(), Map.of());
+            return ResponseEntity.of(request.getPath(), file);
+        }
+        final byte[] file = fileService.serveFile(request.getPath());
+        return ResponseEntity.of(request.getPath(), file);
+    }
+
+    private ResponseEntity getUserList(final RequestEntity request) throws IOException {
+        if (!isLogin(request)) {
+            return ResponseEntity.redirectResponseEntity("/user/login.html");
+        }
+        final Collection<User> users = userService.findAll();
+        final byte[] file = fileService.render(request.getPath(), Map.of("users", users));
+        return ResponseEntity.of(request.getPath(), file);
+    }
+
     private ResponseEntity createUser(final RequestEntity request) {
         if (request.isGet()) {
-            userService.createUser(new CreateUserDto(request.getHeader().getQueryParameter()));
+            validateCreateUser(request.getQueryParameters());
+            userService.createUser(CreateUserDto.of(request.getQueryParameters()));
+            return ResponseEntity.redirectResponseEntity("/index.html");
+        }
+        if (request.isPost()) {
+            validateCreateUser(request.getBody().get());
+            userService.createUser(CreateUserDto.of(request.getBody().get()));
+            return ResponseEntity.redirectResponseEntity("/index.html");
         }
-        userService.createUser(new CreateUserDto(request.getBody().get()));
-        return ResponseEntity.redirectResponseEntity("/index.html");
+
+        throw new IllegalArgumentException("존재하지 않는 요청입니다.");
+    }
+
+    private void validateCreateUser(final Map<String, String> values) {
+        if (values.get("userId") == null ||
+                values.get("password") == null ||
+                values.get("name") == null ||
+                values.get("email") == null) {
+            throw new IllegalArgumentException("입력값이 올바르지 않습니다.");
+        }
+    }
+
+    private ResponseEntity login(final RequestEntity request) {
+        if (request.isPost()) {
+            try {
+                final User user = userService.login(new LoginDto(request.getBody().get("userId"), request.getBody().get("password")));
```

Comment:
httpRequest객체에서 요청 값을  body에서  꺼내야할지 QueryParameter에서 꺼내와야할지가 컨트롤러의 메서드(= 핸들러)의 관심사는 아니라고 생각해요. 🙂

`httpRequest.getParameter("key")`와 같이 httpRequest에게는 단순히 메시지만 보내도록 변경해볼 수 있을 것 같아요. 🙂

```suggestion
                final User user = userService.login(new LoginDto(request.getParameter("userId"), request.getParameter("password")));
```

Reply:
피드백 반영했습니다! https://github.com/next-step/jwp-was-kakao/pull/31/commits/06f6ca2e616e804943f3bdd587d30fd3bf15680d